### PR TITLE
Added active/active test suite

### PR DESCRIPTION
**Pull request by miminar**
Added testing framework for controllers election.

Tests Can be run with:

    ./hack/test-extended.sh ha

In their setup a standalone etcd server, api server and a node are
launched. Individual controllers are started/terminated by ginkgo test
suite (prefixed with `ha:`).

### CODE REVIEW

**Comment by miminar** [comment]
ping @deads2k, @ironcladlou

File: hack/test-extended/ha/run.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+trap "cleanup" EXIT
+
+export TMPDIR="${TMPDIR:-"/tmp"}"
+export BASETMPDIR="${TMPDIR}/openshift-extended-tests/images"
+setup_env_vars
+export MASTER_CONFIG_PATH="${MASTER_CONFIG_DIR}/master-config.yaml"
+reset_tmp_dir 
+
+configure_os_server
+sed -i "s/\(\<controllerLeaseTTL\s*:\s*\)[0-9]\+/\1$CONTROLLER_LEASE_TTL/" \
```
Comment (ironcladlou) [inline hack/test-extended/ha/run.sh]: Are we allowed to start relying on `jq`?

Comment (miminar) [inline hack/test-extended/ha/run.sh]: Would be nice. In this case `yq` would be more appropriate tough :-). It made me realize that I make an assumption about master config being yaml. I guess the override could be passed to `configure_os_server` like this:

    configure_os_server controllerLeaseTTL=XY other.nested.var=YZ 'some.string.var="foo"'

which would handle dirty work. In that case `jq` would really come handy.

**Comment by miminar** [comment]
Rebased.

**Comment by deads2k** [comment]
> Are we allowed to start relying on jq?

If `yq` and `jq` aren't already in our image, I'm in favor of adding them.  @danmcp any objection?

**Comment by deads2k** [comment]
[test] [extended:ha]

File: Godeps/Godeps.json [hunk 2]
```diff
@@ -11,19 +11,10 @@
 			"Rev": "75cd24fc2f2c2a2088577d12123ddee5f54e0675"
 		},
 		{
-			"ImportPath": "code.google.com/p/gcfg",
```
Comment (deads2k) [inline Godeps/Godeps.json]: This seems strange.  I see the `json` file updated, but I don't see any updated files.  Mistaken commit somewhere?

File: assets/test/protractor-chrome.conf.js [hunk 3]
```diff
@@ -101,7 +101,7 @@ exports.config = {
   // testing on multiple browsers, use the multiCapabilities array.
 
   // For a list of available capabilities, see
-  // https://code.google.com/p/selenium/wiki/DesiredCapabilities
+  // https://github.com/SeleniumHQ/selenium/wiki/DesiredCapabilities
```
Comment (deads2k) [inline assets/test/protractor-chrome.conf.js]: I don't think belongs in this commit/pull

File: hack/util.sh [hunk 4]
```diff
@@ -152,6 +152,44 @@ function start_os_server {
 	echo `date`
 }
 
+# start_os_api_server starts standalone OS's API. Node and Controllers need to be started
+# separately. Useful for testing controllers election.
+function start_os_api_server {
```
Comment (deads2k) [inline hack/util.sh]: This is currently specific to your extended test?  Let's keep this function in a file under your extended test.  I'm fine with the separate function, but I'd like to keep the main `util.sh` a "90% case" kind of file and this is pretty specific.

**Comment by danmcp** [comment]
> If `yq` and `jq` aren't already in our image, I'm in favor of adding them.  @danmcp any objection?

@deads2k no

File: hack/util.sh [hunk 5]
```diff
@@ -368,6 +406,64 @@ function validate_response {
 	return 1
 }
 
+# start_etcd starts an etcd server
+# $1 - Optional host (Default: 127.0.0.1)
+# $2 - Optional port (Default: 4001)
+function start_etcd_extended() {
```
Comment (deads2k) [inline hack/util.sh]: Ok, I recognize this function.  It shouldn't be `_extended`, if this can't support the general case of starting etcd, then it belongs in your extended test.  Since this very similar to `test-integration.sh`, I'm ok with having such a function here, but I'd say simplify it.

 1. Use "standard" env vars in this file to start it (similar to starting the openshift server).
 1. Don't inspect currently running processes.  That was probably a mistake in the first script.
 1. You don't need to check etcd version, since we only support v2 now.

Comment (miminar) [inline hack/util.sh]: Ok, will do.

File: hack/util.sh (hunk 4 above)
Comment (miminar) [inline hack/util.sh]: Alright, that's reasonable.

File: assets/test/protractor-chrome.conf.js (hunk 3 above)
Comment (miminar) [inline assets/test/protractor-chrome.conf.js]: I had troubles compiling OS because of code.google.com repos migrating to github and other places so I updated affected imports in the [first commit](https://github.com/miminar/origin/commit/57df7ce2634c3eb5a78951e3542f967d6d3b4403). I guess it should be a separate PR.

File: test/extended/ha/ha.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+	// Run a controller manager with one intitial controllers instance.
+	// Running controllers are needed for NewCLI to run.
+	// FIXME: seems like OutputDir is unset at this moment (execution of the first test)
+	m, _err := exutil.NewControllersManager(exutil.MasterConfigPath(), 0, 8444, exutil.TestContext.OutputDir)
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
+	}
+	// FIXME: Controllers instance immediatelly terminates after acquiring its lease without following sleep
+	time.Sleep(1 * time.Second)
```
Comment (deads2k) [inline test/extended/ha/ha.go]: No.  This needs to get sorted out.  Sleeps are way too fragile and indicate timing problems.  If you know what you need to wait for a `Poll` is probably acceptable, but I'd prefer to fix the problem.

File: test/extended/ha/ha.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	o "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = g.Describe("ha: Election of OpenShift controllers", func() {
+	defer g.GinkgoRecover()
+	var (
+		podConfig  = exutil.FixturePath("fixtures", "hello-world-pod.json")
+		reExitCode = regexp.MustCompile(`(?i)exitcode:(\d+)`)
```
Comment (deads2k) [inline test/extended/ha/ha.go]: There should be another way to get this information.  @mfojtik ?

Also, I don't recognize the string "exitcode".  Is it from our test code?

File: test/extended/util/controllersmanager.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+)
+
+const (
+	EtcdLeasePath = `/openshift.io/leases/controllers`
+)
+
+// ControllersManager allows to start and stop controllers instances,
+// distinguish currently active one by looking up leaseID in etcd storage and
+// allows for updating and deleting of the lease.
+type ControllersManager struct {
```
Comment (deads2k) [inline test/extended/util/controllersmanager.go]: A manager for our managers (controllers) concerns me.  I'll keep reading, but I expected this test to be building and scaling a replication controller for our controllers.  Maybe create a bare pod to prime the system, but then an RC to test scaling out.

File: test/extended/util/framework.go [hunk 9]
```diff
@@ -106,6 +106,24 @@ func WaitForBuilderAccount(c kclient.ServiceAccountsInterface) error {
 	return wait.Poll(60, time.Duration(1*time.Second), waitFunc)
 }
 
+// WaitForDeployerAccount waits until the deployer service account gets fully
+// provisioned
+func WaitForDeployerAccount(c kclient.ServiceAccountsInterface) error {
+	waitFunc := func() (bool, error) {
+		sc, err := c.Get("deployer")
```
Comment (deads2k) [inline test/extended/util/framework.go]: There's a constant for this in the `bootstrappolicy` package.

File: test/extended/util/framework.go [hunk 10]
```diff
@@ -106,6 +106,24 @@ func WaitForBuilderAccount(c kclient.ServiceAccountsInterface) error {
...
+		if err != nil {
+			return false, err
+		}
+		for _, s := range sc.Secrets {
+			if strings.Contains(s.Name, "dockercfg") {
```
Comment (deads2k) [inline test/extended/util/framework.go]: The deployer needs a non-empty list of `ImagePullSecrets`, as opposed to mountable secret containing that value.

File: test/extended/ha/ha.go (hunk 7 above)
Comment (miminar) [inline test/extended/ha/ha.go]: It's for parsing output of `oc get --watch pod/<pod>`. I belive it could be replaced with `oc logs -f`.

File: test/extended/ha/ha.go (hunk 6 above)
Comment (miminar) [inline test/extended/ha/ha.go]: I'll look into this.

File: test/extended/ha/ha.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+	if _err != nil {
+		g.Fail(_err.Error())
+	}
+
+	oc := exutil.NewCLI("controller-leases", exutil.KubeConfigPath())
+
+	// WaitForPod blocks until given pod terminates.
+	// FIXME: this is probably too complicated
+	// FIXME: consider using oc logs directly
+	WaitForPod := func(podName string, timeout time.Duration) {
```
Comment (deads2k) [inline test/extended/ha/ha.go]: This is big enough that I'd like to see it factored as a named function. 

@smarterclayton did your sync command ever go in?

File: test/extended/ha/ha.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+				cmd.Wait()
+			} else {
+				o.Expect(cmd.Wait()).NotTo(o.HaveOccurred())
+			}
+		case <-time.After(start.Add(timeout).Sub(time.Now())):
+			g.Fail(fmt.Sprintf("Timeout (%s) occurred while waiting for pod's exit", timeout.String()))
+		}
+	}
+
+	StartPod := func() {
```
Comment (deads2k) [inline test/extended/ha/ha.go]: Explain what this is for.  Is this just a function to ensure that the server still works?  If so, I'd like to see a deploymentconfig.

File: test/extended/ha/ha.go [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+
+	// EnsureOneInstanceActive releases any inactive controllers instances and
+	// makes sure that one is active.
+	EnsureOneInstanceActive := func() {
+		l := m.Len()
+		switch {
+		case l > 1:
+			g.By("\n HA TEST:  Releasing inactive controllers")
+			m.ReleaseControllers(m.GetInactive()...)
+		case l == 0:
```
Comment (deads2k) [inline test/extended/ha/ha.go]: don't use `l`.  It led to me scratching my head saying, "wtf, why is he testing if one equals zero??" :)

File: test/extended/ha/ha.go (hunk 11 above)
Comment (smarterclayton) [inline test/extended/ha/ha.go]: No, still have to sort through the experience

On Sep 15, 2015, at 12:30 PM, David Eads <notifications@github.com> wrote:

In test/extended/ha/ha.go
<https://github.com/openshift/origin/pull/4628#discussion_r39532364>:

> +	// One controllers instance needs to be running before CLI can be initialized
> +	_, _err = m.StartNewInstance()
> +	if _err != nil {
> +		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
> +	}
> +	_, _, _err = m.WaitForActive(time.Second * 2)
> +	if _err != nil {
> +		g.Fail(_err.Error())
> +	}
> +
> +	oc := exutil.NewCLI("controller-leases", exutil.KubeConfigPath())
> +
> +	// WaitForPod blocks until given pod terminates.
> +	// FIXME: this is probably too complicated
> +	// FIXME: consider using oc logs directly
> +	WaitForPod := func(podName string, timeout time.Duration) {

This is big enough that I'd like to see it factored as a named function.

@smarterclayton <https://github.com/smarterclayton> did your sync command
ever go in?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4628/files#r39532364>.

**Comment by deads2k** [comment]
In general, what problems do you see with trying to test this using a replication controller with a host-mount volume?

**Comment by smarterclayton** [comment]
Do we need a host mount volume?  Why not use a secret?  (maybe I missed
something in my reading of the pull)

On Sep 15, 2015, at 12:37 PM, David Eads <notifications@github.com> wrote:

In general, what problems do you see with trying to test this using a
replication controller with a host-mount volume?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4628#issuecomment-140455821>.

File: test/extended/ha/ha.go [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
+	}
+	// FIXME: Controllers instance immediatelly terminates after acquiring its lease without following sleep
+	time.Sleep(1 * time.Second)
+
+	// One controllers instance needs to be running before CLI can be initialized
+	_, _err = m.StartNewInstance()
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
```
Comment (mfojtik) [inline test/extended/ha/ha.go]: o.Expect(err).NotTo(o.HaveOccurred()) ?

File: test/extended/ha/ha.go [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+	}
+	_, _, _err = m.WaitForActive(time.Second * 2)
+	if _err != nil {
+		g.Fail(_err.Error())
```
Comment (mfojtik) [inline test/extended/ha/ha.go]: o.Expect(err).NotTo(o.HaveOccurred()) ?

File: test/extended/ha/ha.go [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+		g.By("\n HA TEST:  waiting for deployer service account")
+		err := exutil.WaitForDeployerAccount(oc.KubeREST().ServiceAccounts(oc.Namespace()))
+		o.Expect(err).NotTo(o.HaveOccurred())
+	})
+
+	g.AfterEach(func() {
+		EnsureOneInstanceActive()
+	})
+
+	g.Describe("\n HA TEST:  Delete an active lease", func() {
```
Comment (mfojtik) [inline test/extended/ha/ha.go]: Why a new line?

Comment (miminar) [inline test/extended/ha/ha.go]: I'm not sure myself. I just wanted to be consistent with another non-default test [forcepull](https://github.com/miminar/origin/blob/test-active-active/test/extended/builds/forcepull.go#L69). @gabemontero ^^? But since other non-default test `networking` doesn't use these prefixes, I think they can be removed.

Comment (mfojtik) [inline test/extended/ha/ha.go]: Yeah, forcepull has couple problems. It uses non-consistent messages and the beforeSuite runs for **every** test (i'm trying to fix ths now)

File: test/extended/ha/ha.go (hunk 13 above)
Comment (soltysh) [inline test/extended/ha/ha.go]: I don't understand why doing a switch in a situation where plain old if-else is sufficient?

File: test/extended/ha/ha.go (hunk 16 above)
Comment (gabemontero) [inline test/extended/ha/ha.go]: On the messages, I wanted an easier way to be able to filter for the messages from the forcepull test if I ran the whole set of tests (I did not find the ginkgo facilities sufficient).  As you can see, not everyone agrees with my opinion on this ;-).

On BeforeSuite, according to the doc at http://onsi.github.io/ginkgo/#global-setup-and-teardown-beforesuite-and-aftersuite,
it only runs "once before the entire test suite".  And that was my observation when developing.  Not sure what @mfojtik means by "runs for every test".

I do see that the focus string for forcepull was changed to "builds: serial:" since then.  I don't know yet if that has any bearing on the forcepull BeforeSuite behavior (I had not tried extended tests since their latest refactoring and I'm still trying to get them to run as of this writing).

Comment (mfojtik) [inline test/extended/ha/ha.go]: @gabemontero beforeSuite runs even when I'm not executing the forcepull tests :-) I can see that from the output. Try running the images test :)

Comment (gabemontero) [inline test/extended/ha/ha.go]: Yep, got the tests running and see that now as well ... I'll create a separate PR for that.  I can ref that PR here if folks are interested.

File: test/extended/ha/ha.go (hunk 14 above)
Comment (miminar) [inline test/extended/ha/ha.go]: It seems that Ω can't be used outside `(Before|After)Each|It`:

    Test Panicked
    You are trying to make an assertion, but Gomega's fail handler is nil.
    If you're using Ginkgo then you probably forgot to put your assertion in an It().
    Alternatively, you may have forgotten to register a fail handler with RegisterFailHandler() or RegisterTestingT().

File: test/extended/ha/ha.go (hunk 16 above)
Comment (smarterclayton) [inline test/extended/ha/ha.go]: It's ok to have focus prefixes but be sure to make them more descriptive
rather than more terse

On Sep 17, 2015, at 9:23 AM, Gabe Montero <notifications@github.com> wrote:

In test/extended/ha/ha.go
<https://github.com/openshift/origin/pull/4628#discussion_r39743927>:

> +	}
> +
> +	g.JustBeforeEach(func() {
> +		oc.SetOutputDir(exutil.TestContext.OutputDir)
> +
> +		o.Expect(m.Len()).To(o.Equal(1))
> +		g.By("\n HA TEST:  waiting for deployer service account")
> +		err := exutil.WaitForDeployerAccount(oc.KubeREST().ServiceAccounts(oc.Namespace()))
> +		o.Expect(err).NotTo(o.HaveOccurred())
> +	})
> +
> +	g.AfterEach(func() {
> +		EnsureOneInstanceActive()
> +	})
> +
> +	g.Describe("\n HA TEST:  Delete an active lease", func() {

On the messages, I wanted an easier way to be able to filter for the
messages from the forcepull test if I ran the whole set of tests (I did not
find the ginkgo facilities sufficient). As you can see, not everyone agrees
with my opinion on this ;-).

On BeforeSuite, according to the doc at
http://onsi.github.io/ginkgo/#global-setup-and-teardown-beforesuite-and-aftersuite
,
it only runs "once before the entire test suite". And that was my
observation when developing. Not sure what @mfojtik
<https://github.com/mfojtik> means by "runs for every test".

I do see that the focus string for forcepull was changed to "builds:
serial:" since then. I don't know yet if that has any bearing on the
forcepull BeforeSuite behavior (I had not tried extended tests since their
latest refactoring and I'm still trying to get them to run as of this
writing).

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4628/files#r39743927>.

File: test/extended/ha/ha.go (hunk 14 above)
Comment (miminar) [inline test/extended/ha/ha.go]: Moved to `g.BeforeSuite()` where `o.Expect(err).NotTo(o.HaveOccurred())` can be run without a problem.

File: test/extended/util/cmonitor.go [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+		index := uint64(0)
+		if latestResp != nil {
+			index = latestResp.Node.ModifiedIndex + 1
+		}
+		go func() {
+			response, err := m.EtcdClient.Watch(EtcdLeasePath, index, false, nil, stopChan)
+			if err != nil {
+				errChan <- err
+			}
+			responseChan <- response
```
Comment (kargakis) [inline test/extended/util/cmonitor.go]: This will block if there is an error because you receive only once below and the goroutine will leak. Do you care about responses in case of an error? Looking at your select below, I don't think so. Return the goroutine if err != nil

File: test/extended/util/cmonitor.go [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+		select {
+		case err := <-errChan:
+			return nil, time.Now().Sub(start), err
+		case resp := <-responseChan:
+			if ctrls := findInstance(resp.Node.Value); ctrls != nil {
+				return ctrls, time.Now().Sub(start), nil
+			}
+			latestResp = resp
+		case <-time.After(start.Add(timeout).Sub(time.Now())):
+			stopChan <- true
```
Comment (kargakis) [inline test/extended/util/cmonitor.go]: stopchans are supposed to broadcast so you should close this instead. I don't see it used anywhere though.

File: test/extended/util/cmonitor.go (hunk 17 above)
Comment (miminar) [inline test/extended/util/cmonitor.go]: Good catch!

File: test/extended/util/cmonitor.go (hunk 18 above)
Comment (miminar) [inline test/extended/util/cmonitor.go]: Should be fixed.

**Comment by miminar** [comment]
@deads2k

> If yq and jq aren't already in our image, I'm in favor of adding them. @danmcp any objection?

Do you mean `openshift/origin` image (`images/origin/Dockerfile`). Any other place that needs to be amended?

> In general, what problems do you see with trying to test this using a replication controller with a host-mount volume?

Won't encapsulation of processes hide possible interferences? I guess it would make the test a bit simpler though.

**Comment by deads2k** [comment]
> Do you mean openshift/origin image (images/origin/Dockerfile). Any other place that needs to be amended?

Given what you're modifying, I'd expect it to be the jenkins ami.

**Comment by deads2k** [comment]
> Won't encapsulation of processes hide possible interferences? I guess it would make the test a bit simpler though.

I think it would work out well, using endpoint inspection, scale, `secrets` to hold the config, and a good readiness check.  It would also make this test operate in a way that helps us describe how to use openshift to run our controllers in a maintainable way.

**Comment by miminar** [comment]
Sounds fair. I'll rework it then.

File: test/extended/util/cmonitor.go (hunk 18 above)
Comment (kargakis) [inline test/extended/util/cmonitor.go]: Also, since you don't care about the actual value of the channel, it should be an empty struct.

**Comment by danmcp** [comment]
@miminar Needs rebase.

**Comment by miminar** [comment]
I'll submit reworked test suite. Closing this.
